_posts/2021-09-06-terraform-uncommon-options.markdown

```diff
@@ -0,0 +1,336 @@
+local_file.private_key: Creating...
+local_file.private_key: Creation complete after 0s [id=e7ec4a8f2309bdd4c4c57cb2adfb79c91a293597]
+
+Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
+```
+
+As expected, Terraform reverts back to the state as instructed by the configuration in the `main.tf.` template.
+
+## [JSON Instead Of HCL]
+The use case for this capability is quite interesting. I can programmatically generate Terraform templates using the JSON syntax. The only requirement is for the filename to have the extension `tf.json` otherwise Terraform will not pick the configuration up.
+I can convert the example above to JSON like so:
+
+```json
+{
+  "locals": {
+    "map": {
+    "a": 1,
+    "b": 2,
+    "c": 3
+    }
+  },
+  "resource": {
+    "local_file": {
+      "private_key": {
+        "content": "jsonencode(local.map)",
+        "filename": "./map.json"
+      }
+    }
+  }
+}
+```
+I would reckon it’s unusual to generate a JSON encoded file using the JSON syntax. It’s a very appealing feature I imagine is being actively used out there.
+
+## [Terraform Graph]
+
+Not the most widely used Terraform command, but most definitely uncommon. It generates a DOT (graph description language) file that I can use to render an image in a well-known format, such as PNG.
+Running this command in the example state yields a chart description configuration.
+
+```sh
+$ terraform graph
+```
+
+```dot
+digraph {
+	compound = "true"
+	newrank = "true"
+	subgraph "root" {
+		"[root] local_file.private_key (expand)" [label = "local_file.private_key", shape = "box"]
+		"[root] provider[\"registry.terraform.io/hashicorp/local\"]" [label = "provider[\"registry.terraform.io/hashicorp/local\"]", shape = "diamond"]
+		"[root] local_file.private_key (expand)" -> "[root] local.map (expand)"
+		"[root] local_file.private_key (expand)" -> "[root] provider[\"registry.terraform.io/hashicorp/local\"]"
+		"[root] meta.count-boundary (EachMode fixup)" -> "[root] local_file.private_key (expand)"
+		"[root] provider[\"registry.terraform.io/hashicorp/local\"] (close)" -> "[root] local_file.private_key (expand)"
+		"[root] root" -> "[root] meta.count-boundary (EachMode fixup)"
+		"[root] root" -> "[root] provider[\"registry.terraform.io/hashicorp/local\"] (close)"
+	}
+}
+```
+
+If I "pipe" this through a filter for drawing directed graphs:
+
+```sh
+terraform graph | dot -Tpng > graph.png
+```
+I get a nice image of a graph as seen in Figure 1. (I've also inverted the colors with `imagemagick`)
+
+<p align="center">
+  <img style="border:1px solid #2C4D56" alt="Graph created from a dot template" title="Graph created from a dot template" src="assets/images/2021_09_dot.png" height="300">
+  <b style="font-size:0.7vw" >Figure 1. Graph created from a DOT template.</b>
+</p>
+<br/>
+
+Getting to know a specific technology or a tool is not as easy as it seems. Hopefully, this blog post will make some of terraform unusual options clearer and more usable.
```

Comment:
the second sentence in this line needs to change to "Hopefully, this blog post will make some of the unusual Terraform options clearer and more usable."

Reply:
I think I wanted to write "Terraform's" here, but your suggestion also works.